src/format.rs

```diff
@@ -0,0 +1,888 @@
+impl UtcOffset {
+    fn new(hour: f64, minute: u32, second: u32) -> Self {
+        Self {
+            hour,
+            minute,
+            second,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Piece {
+    width: Option<usize>,
+    padding: Padding,
+    has_left_padding: bool,
+    case: Option<Case>,
+    spec: Spec,
+}
+
+impl Piece {
+    fn new(
+        width: Option<usize>,
+        padding: Padding,
+        has_left_padding: bool,
+        case: Option<Case>,
+        spec: Spec,
+    ) -> Self {
+        Self {
+            width,
+            padding,
+            has_left_padding,
+            case,
+            spec,
+        }
+    }
+
+    fn format_num_zeros(
+        &self,
+        f: &mut SizeLimiter<'_>,
+        value: impl fmt::Display,
+        default_width: usize,
+    ) -> io::Result<()> {
+        if self.has_left_padding {
+            write!(f, "{}", value)
+        } else if self.padding == Padding::Spaces {
+            let width = self.width.unwrap_or(default_width);
+            write!(f, "{: >width$}", value)
+        } else {
+            let width = self.width.unwrap_or(default_width);
+            write!(f, "{:0width$}", value)
+        }
+    }
+
+    fn format_num_spaces(
+        &self,
+        f: &mut SizeLimiter<'_>,
+        value: impl fmt::Display,
+        default_width: usize,
+    ) -> io::Result<()> {
+        if self.has_left_padding {
+            write!(f, "{}", value)
+        } else if self.padding == Padding::Zeros {
+            let width = self.width.unwrap_or(default_width);
+            write!(f, "{:0width$}", value)
+        } else {
+            let width = self.width.unwrap_or(default_width);
+            write!(f, "{: >width$}", value)
+        }
+    }
+
+    fn format_nanoseconds(
+        &self,
+        f: &mut SizeLimiter<'_>,
+        nanoseconds: u32,
+        default_width: usize,
+    ) -> io::Result<()> {
+        let width = self.width.unwrap_or(default_width);
+
+        if width <= 9 {
+            let value = nanoseconds / 10u32.pow(9 - width as u32);
+            write!(f, "{:0n$}", value, n = width)
+        } else {
+            write!(f, "{:09}{:0n$}", nanoseconds, 0, n = width - 9)
+        }
+    }
+
+    fn format_string(&self, f: &mut SizeLimiter<'_>, s: impl fmt::Display) -> io::Result<()> {
+        match self.width {
+            None => write!(f, "{}", s),
+            Some(width) => {
+                if self.has_left_padding {
+                    write!(f, "{}", s)
+                } else if self.padding == Padding::Zeros {
+                    write!(f, "{:0>width$}", s)
+                } else {
+                    write!(f, "{: >width$}", s)
+                }
+            }
+        }
+    }
+
+    fn compute_offset_parts(&self, time: &Time) -> UtcOffset {
+        let utc_offset = time.utc_offset();
+        let utc_offset_abs = utc_offset.unsigned_abs();
+
+        let sign = if utc_offset < 0 || time.is_utc() && self.has_left_padding {
```

Comment:
This won't work, since the position of the sign depends of the type of padding:
```
irb(main):015:0> Time.now.utc.strftime("%-_10z")
=> "      -000"
irb(main):016:0> Time.now.utc.strftime("%-10z")
=> "-000000000"
```

This is handled the using the special [`+` and `0` specifiers](https://doc.rust-lang.org/std/fmt/index.html#sign0) for the `f64` type:
```rust
write!(f, "{: >+n$.0}", hour); // padding with spaces
write!(f, "{:+0n$.0}", hour); // padding with zeros
```